File: app/src/main/java/org/connectbot/util/Iterm2ColorSchemeParser.java
/*
 * ConnectBot: simple, powerful, open-source SSH client for Android
 * Copyright 2026
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.util;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import java.io.StringReader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilderFactory;

/**
 * Parses iTerm2 {@code .itermcolors} files (plist XML) into a 16-color ANSI palette.
 *
 * <p>iTerm2 schemes are XML plists where ANSI colors are stored under keys like:
 * {@code Ansi 0 Color} .. {@code Ansi 15 Color}.</p>
 *
 * <p>Optional keys like {@code Foreground Color} / {@code Background Color} may exist and are
 * returned (as ARGB ints) for optional default-FG/BG heuristics.</p>
 */
public final class Iterm2ColorSchemeParser {
	private Iterm2ColorSchemeParser() {
	}

	public static final class ParsedScheme {
		public final int[] ansiColors;
		public final Integer foregroundColor;
		public final Integer backgroundColor;

		private ParsedScheme(int[] ansiColors, Integer foregroundColor, Integer backgroundColor) {
			this.ansiColors = ansiColors;
			this.foregroundColor = foregroundColor;
			this.backgroundColor = backgroundColor;
		}
	}

	private static final Pattern ANSI_KEY = Pattern.compile("^Ansi (\\d+) Color$");

	public static boolean looksLikeIterm2Scheme(String text) {
		if (text == null) return false;
		String trimmed = ltrim(text);
		return trimmed.startsWith("<") && trimmed.toLowerCase().contains("<plist");
	}

	public static ParsedScheme parse(String xmlPlist) {
		if (xmlPlist == null) {
			throw new IllegalArgumentException("xmlPlist is null");
		}
		String trimmed = ltrim(xmlPlist);
		if (!trimmed.startsWith("<") || trimmed.toLowerCase().indexOf("<plist") < 0) {
			throw new IllegalArgumentException("Not an iTerm2 plist");
		}

		try {
			DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
			factory.setNamespaceAware(false);
			factory.setXIncludeAware(false);
			factory.setExpandEntityReferences(false);

			// The common iTerm2 plist includes a DOCTYPE referencing Apple's DTD.
			// Avoid fetching external resources (XXE / network fetches).
			trySetFeature(factory, XMLConstants.FEATURE_SECURE_PROCESSING, true);
			trySetFeature(factory, "http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
			trySetFeature(factory, "http://xml.org/sax/features/external-general-entities", false);
			trySetFeature(factory, "http://xml.org/sax/features/external-parameter-entities", false);

			javax.xml.parsers.DocumentBuilder builder = factory.newDocumentBuilder();
			builder.setEntityResolver((publicId, systemId) -> new InputSource(new StringReader("")));

			org.w3c.dom.Document document = builder.parse(new InputSource(new StringReader(xmlPlist)));
			Element root = document.getDocumentElement();
			if (root == null || !"plist".equals(root.getTagName())) {
				throw new IllegalArgumentException("Invalid plist root");
			}

			Element outerDict = firstChildElement(root, "dict");
			if (outerDict == null) {
				throw new IllegalArgumentException("Missing plist dict");
			}

			int[] palette = new int[16];
			System.arraycopy(Colors.defaults, 0, palette, 0, 16);
			boolean[] foundAnsi = new boolean[16];
			final Integer[] fgBg = new Integer[] {null, null};

			forEachPlistKeyValue(outerDict, new KeyValueVisitor() {
				@Override
				public void visit(String key, Element valueElement) {
					if (key == null) return;
					Matcher m = ANSI_KEY.matcher(key);
					if (m.matches()) {
						int idx;
						try {
							idx = Integer.parseInt(m.group(1));
						} catch (NumberFormatException e) {
							return;
						}
						if (idx >= 0 && idx < 16) {
							int color = parseColorDict(valueElement);
							palette[idx] = color;
							foundAnsi[idx] = true;
						}
					} else if ("Foreground Color".equals(key)) {
						fgBg[0] = parseColorDict(valueElement);
					} else if ("Background Color".equals(key)) {
						fgBg[1] = parseColorDict(valueElement);
					}
				}
			});

			boolean any = false;
			for (boolean b : foundAnsi) {
				if (b) {
					any = true;
					break;
				}
			}
			if (!any) {
				throw new IllegalArgumentException("No ANSI colors found in plist");
			}
			for (int i = 0; i < foundAnsi.length; i++) {
				if (!foundAnsi[i]) {
					throw new IllegalArgumentException("Missing ANSI colors: " + i);
				}
			}

			return new ParsedScheme(palette, fgBg[0], fgBg[1]);
		} catch (IllegalArgumentException e) {
			throw e;
		} catch (Exception e) {
			throw new IllegalArgumentException("Failed to parse plist: " + e.getMessage(), e);
		}
	}

	private static void trySetFeature(DocumentBuilderFactory factory, String feature, boolean value) {
		try {
			factory.setFeature(feature, value);
		} catch (Exception ignored) {
			// Best-effort: parser implementations vary across Android/JVM.
		}
	}

	private static String ltrim(String s) {
		int i = 0;
		while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
		return s.substring(i);
	}

	private static Element firstChildElement(Element parent, String tagName) {
		NodeList nodes = parent.getChildNodes();
		for (int i = 0; i < nodes.getLength(); i++) {
			Node node = nodes.item(i);
			if (node instanceof Element) {
				Element e = (Element) node;
				if (tagName.equals(e.getTagName())) {
					return e;
				}
			}
		}
		return null;
	}

	private interface KeyValueVisitor {
		void visit(String key, Element valueElement);
	}

	private static void forEachPlistKeyValue(Element dict, KeyValueVisitor visitor) {
		if (!"dict".equals(dict.getTagName())) {
			throw new IllegalArgumentException("Expected dict");
		}
		NodeList children = dict.getChildNodes();
		for (int i = 0; i < children.getLength(); i++) {
			Node node = children.item(i);
			if (node instanceof Element && "key".equals(((Element) node).getTagName())) {
				String key = ((Element) node).getTextContent().trim();
				Element value = nextElementSibling(children, i + 1);
				if (value == null) {
					throw new IllegalArgumentException("Missing value for key: " + key);
				}
				visitor.visit(key, value);
			}
		}
	}

	private static Element nextElementSibling(NodeList nodes, int startIndex) {
		for (int i = startIndex; i < nodes.getLength(); i++) {
			Node node = nodes.item(i);
			if (node instanceof Element) {
				return (Element) node;
			}
		}
		return null;
	}

	private static int parseColorDict(Element colorValueElement) {
		if (!"dict".equals(colorValueElement.getTagName())) {
			throw new IllegalArgumentException("Color value is not a dict");
		}

		final double[] rgba = new double[] {0.0, 0.0, 0.0, 1.0};

		forEachPlistKeyValue(colorValueElement, (key, value) -> {
			String text = value.getTextContent();
			if (text == null) return;
			double numeric;
			try {
				numeric = Double.parseDouble(text.trim());
			} catch (NumberFormatException e) {
				return;
			}
			switch (key) {
				case "Red Component":
					rgba[0] = numeric;
					break;
				case "Green Component":
					rgba[1] = numeric;
					break;
				case "Blue Component":
					rgba[2] = numeric;
					break;
				case "Alpha Component":
					rgba[3] = numeric;
					break;
				default:
					break;
			}
		});

		int r = toByte(normalizeComponent(rgba[0]));
		int g = toByte(normalizeComponent(rgba[1]));
		int b = toByte(normalizeComponent(rgba[2]));
		int a = toByte(normalizeComponent(rgba[3]));
		return (a << 24) | (r << 16) | (g << 8) | b;
	}

	private static double normalizeComponent(double value) {
		if (Double.isNaN(value)) return 0.0;
		if (value <= 0.0) return 0.0;
		if (value <= 1.0) return value;
		if (value <= 255.0) return value / 255.0;
		return 1.0;
	}

	private static int toByte(double v) {
		int b = (int) Math.round(v * 255.0);
		if (b < 0) return 0;
		if (b > 255) return 255;
		return b;
	}
}

File: app/src/main/java/org/connectbot/data/Iterm2ColorSchemeParser.kt
/*
 * ConnectBot: simple, powerful, open-source SSH client for Android
 * Copyright 2026
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.data

import org.w3c.dom.Element
import java.io.StringReader
import javax.xml.XMLConstants
import javax.xml.parsers.DocumentBuilderFactory
import kotlin.math.roundToInt
import org.xml.sax.InputSource

/**
 * Parses iTerm2 `.itermcolors` files (plist XML) into a terminal palette.
 *
 * iTerm2 schemes are XML plists where ANSI colors are stored under keys like:
 * `Ansi 0 Color` .. `Ansi 15 Color`.
 *
 * Additional keys like `Foreground Color` / `Background Color` may exist and are
 * returned (as ARGB ints) for optional default-FG/BG heuristics.
 */
object Iterm2ColorSchemeParser {

    data class ParsedScheme(
        val ansiColors: IntArray,
        val foregroundColor: Int?,
        val backgroundColor: Int?,
    )

    fun parse(xmlPlist: String): ParsedScheme {
        val trimmed = xmlPlist.trimStart()
        require(trimmed.startsWith("<") && trimmed.contains("<plist")) {
            "Not an iTerm2 plist"
        }

        val documentBuilder = secureDocumentBuilder()
        val document = documentBuilder.parse(InputSource(StringReader(xmlPlist)))
        val root = document.documentElement
        require(root != null && root.tagName == "plist") { "Invalid plist root" }

        val outerDict = firstChildElement(root, "dict") ?: throw IllegalArgumentException("Missing plist dict")

        val palette = IntArray(16) { ColorSchemePresets.default.colors[it] }
        val foundAnsi = BooleanArray(16)

        var foreground: Int? = null
        var background: Int? = null

        forEachPlistKeyValue(outerDict) { key, valueElement ->
            when {
                key.startsWith("Ansi ") && key.endsWith(" Color") -> {
                    val index = ANSI_KEY_REGEX.matchEntire(key)?.groupValues?.get(1)?.toIntOrNull()
                    if (index != null && index in 0..15) {
                        val color = parseColorDict(valueElement)
                        palette[index] = color
                        foundAnsi[index] = true
                    }
                }
                key == "Foreground Color" -> foreground = parseColorDict(valueElement)
                key == "Background Color" -> background = parseColorDict(valueElement)
            }
        }

        if (foundAnsi.none { it }) {
            throw IllegalArgumentException("No ANSI colors found in plist")
        }
        if (foundAnsi.any { !it }) {
            val missing = mutableListOf<Int>()
            for (i in foundAnsi.indices) {
                if (!foundAnsi[i]) missing.add(i)
            }
            throw IllegalArgumentException("Missing ANSI colors: $missing")
        }

        return ParsedScheme(
            ansiColors = palette,
            foregroundColor = foreground,
            backgroundColor = background,
        )
    }

    private val ANSI_KEY_REGEX = Regex("""^Ansi (\d+) Color$""")

    private fun secureDocumentBuilder() =
        DocumentBuilderFactory.newInstance().apply {
            isNamespaceAware = false
            isXIncludeAware = false
            setExpandEntityReferences(false)

            // The common iTerm2 plist includes a DOCTYPE referencing Apple's DTD.
            // We must avoid fetching external resources.
            trySetFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true)
            trySetFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
            trySetFeature("http://xml.org/sax/features/external-general-entities", false)
            trySetFeature("http://xml.org/sax/features/external-parameter-entities", false)
        }.newDocumentBuilder().apply {
            setEntityResolver { _, _ -> InputSource(StringReader("")) }
        }

    private fun DocumentBuilderFactory.trySetFeature(feature: String, value: Boolean) {
        try {
            setFeature(feature, value)
        } catch (_: Exception) {
            // Best-effort: parser implementations vary across Android/JVM.
        }
    }

    private fun firstChildElement(parent: Element, tagName: String): Element? {
        val nodes = parent.childNodes ?: return null
        for (i in 0 until nodes.length) {
            val node = nodes.item(i)
            if (node is Element && node.tagName == tagName) return node
        }
        return null
    }

    private fun nextElementSibling(nodes: org.w3c.dom.NodeList, startIndex: Int): Element? {
        for (i in startIndex until nodes.length) {
            val node = nodes.item(i)
            if (node is Element) return node
        }
        return null
    }

    private inline fun forEachPlistKeyValue(dict: Element, block: (key: String, value: Element) -> Unit) {
        require(dict.tagName == "dict") { "Expected dict" }
        val children = dict.childNodes
        var i = 0
        while (i < children.length) {
            val node = children.item(i)
            if (node is Element && node.tagName == "key") {
                val key = node.textContent.trim()
                val valueElement = nextElementSibling(children, i + 1)
                    ?: throw IllegalArgumentException("Missing value for key: $key")
                block(key, valueElement)
            }
            i++
        }
    }

    private fun parseColorDict(colorValueElement: Element): Int {
        require(colorValueElement.tagName == "dict") { "Color value is not a dict" }

        var red = 0.0
        var green = 0.0
        var blue = 0.0
        var alpha = 1.0

        forEachPlistKeyValue(colorValueElement) { key, value ->
            val numericValue = value.textContent.trim().toDoubleOrNull() ?: return@forEachPlistKeyValue
            when (key) {
                "Red Component" -> red = numericValue
                "Green Component" -> green = numericValue
                "Blue Component" -> blue = numericValue
                "Alpha Component" -> alpha = numericValue
            }
        }

        return argbInt(
            r = normalizeComponent(red),
            g = normalizeComponent(green),
            b = normalizeComponent(blue),
            a = normalizeComponent(alpha),
        )
    }

    private fun normalizeComponent(value: Double): Double =
        when {
            value.isNaN() -> 0.0
            value <= 0.0 -> 0.0
            value <= 1.0 -> value
            value <= 255.0 -> value / 255.0
            else -> 1.0
        }

    private fun argbInt(r: Double, g: Double, b: Double, a: Double): Int {
        val rr = (r * 255.0).roundToInt().coerceIn(0, 255)
        val gg = (g * 255.0).roundToInt().coerceIn(0, 255)
        val bb = (b * 255.0).roundToInt().coerceIn(0, 255)
        val aa = (a * 255.0).roundToInt().coerceIn(0, 255)
        return (aa shl 24) or (rr shl 16) or (gg shl 8) or bb
    }
}
